src/dbt_client/dbtClient.ts

```diff
@@ -0,0 +1,124 @@
+import { OutputChannel, window } from 'vscode';
+import { OnSourceFileChanged, SourceFileChangedEvent } from '../manifest/sourceFileChangedEvent';
+import { CommandProcessExecution } from '../utils';
+import { dbtClientCommandQueue as dbtClientCommandQueue } from './dbtClientCommandQueue';
+
+enum DBT {
+	Install,
+	Update
+}
+
+enum PromptAnswer {
+	Yes = "Yes",
+	No = "No"
+}
+
+class DBTClient implements OnSourceFileChanged {
+	public pythonPath!: string;
+	static readonly IS_INSTALLED_VERSION = /(?<=installed\sversion:\s)(\d+.\d+.\d+)(?=\D+)/g;
+	static readonly IS_LATEST_VERSION = /(?<=latest\sversion:\s)(\d+.\d+.\d+)(?=\D+)/g;
+	static readonly IS_INSTALLED = /installed\sversion/g;
+	public outputChannel: OutputChannel;
+
+	constructor() {
+		this.outputChannel = window.createOutputChannel('DBT Client');
+	}
+
+	async onSourceFileChanged(event: SourceFileChangedEvent): Promise<void> {
+		await this.DBTListCommandAndShowOutput(event.projectRoot.fsPath);
+	}
+
+	public async checkDBTInstalled(): Promise<void> {
+		const checkDBTInstalledProcess = this.checkDBTInstalledCommand();
+		try {
+			await checkDBTInstalledProcess.complete();
+		} catch (err) {
+			if (err.match(DBTClient.IS_INSTALLED)) {
+				const DBTUpToDate = this.checkDBTUpToDate(err);
+				if (DBTUpToDate === false) {
+					await this.tryToOperate(DBT.Update);
+					return;
+				}
+				return;
+			}
+			await this.tryToOperate(DBT.Install);
+		}
+	}
+
+	public DBTRunModelCommand(plusOperatorLeft: string, modelName: string, plusOperatorRight: string, cwd: string) {
+		return new CommandProcessExecution(this.pythonPath, ["-c", this.dbtCommand(["'run'", "'--model'", `"${plusOperatorLeft}"`, `"${modelName}"`, `"${plusOperatorRight}"`])], cwd);
+	}
+
+	private DBTListCommand(cwd: string): CommandProcessExecution {
+		return new CommandProcessExecution(this.pythonPath, ["-c", this.dbtCommand("'list'")], cwd);
+	}
+
+	public async DBTListCommandAndShowOutput(cwd: string): Promise<void> {
+		dbtClientCommandQueue.addToQueue(() => this.DBTListCommand(cwd).completeWithOutputChannel(this.outputChannel));
+	}
+
+	private checkDBTInstalledCommand(): CommandProcessExecution {
+		return new CommandProcessExecution(this.pythonPath, ["-c", this.dbtCommand("'--version'")]);
+	}
+
+	private dbtCommand(cmd: string | string[]): string {
+		return `import dbt.main; dbt.main.main([${cmd}])`;
+	}
+
+	private checkDBTUpToDate(message: string): boolean | undefined {
+		const installedVersionMatch = message.match(DBTClient.IS_INSTALLED_VERSION);
+		if (installedVersionMatch === null) {
+			return;
+		}
+		const installedVersion = installedVersionMatch[0];
+		const latestVersionMatch = message.match(DBTClient.IS_LATEST_VERSION);
+		if (latestVersionMatch === null) {
+			return;
+		}
+		const latestVersion = latestVersionMatch[0];
+		if (installedVersion === latestVersion) {
```

Comment:
convoluted logic..

return installedVersion === latestVersion does the same as your 3 lines